explorer-backend/fly/transactions/alephium.go

```diff
@@ -0,0 +1,281 @@
+package transactions
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	sdk "github.com/alephium/go-sdk"
+	"github.com/alephium/wormhole-fork/node/pkg/alephium"
+	"github.com/alephium/wormhole-fork/node/pkg/common"
+	"github.com/alephium/wormhole-fork/node/pkg/vaa"
+	"github.com/btcsuite/btcutil/base58"
+	"go.uber.org/zap"
+)
+
+const ALPHContractAddressPrefix byte = 3
+
+type AlephiumWatcher struct {
+	client             *alephium.Client
+	chainConfig        *common.ChainConfig
+	logger             *zap.Logger
+	blockTxsC          chan<- []*BlockTransactions
+	explorerBackendUrl string
+	fromEventIndex     uint32
+	pollInterval       uint
+}
+
+func NewAlephiumWatcher(
+	config *common.ChainConfig,
+	nodeUrl string,
+	apiKey string,
+	logger *zap.Logger,
+	blockTxsC chan<- []*BlockTransactions,
+	explorerBackendUrl string,
+	fromEventIndex uint32,
+	pollInterval uint,
+) *AlephiumWatcher {
+	client := alephium.NewClient(nodeUrl, apiKey, 10)
+	return &AlephiumWatcher{
+		client:             client,
+		chainConfig:        config,
+		logger:             logger,
+		blockTxsC:          blockTxsC,
+		explorerBackendUrl: explorerBackendUrl,
+		fromEventIndex:     fromEventIndex,
+		pollInterval:       pollInterval,
+	}
+}
+
+func (w *AlephiumWatcher) Run(ctx context.Context, errC chan error) {
+	nodeInfo, err := w.client.GetNodeInfo(ctx)
+	if err != nil {
+		w.logger.Error("failed to get node info", zap.Error(err))
+		errC <- err
+		return
+	}
+
+	isSynced, err := w.client.IsCliqueSynced(ctx)
+	if err != nil {
+		w.logger.Error("failed to get self cliqued synced", zap.Error(err))
+		errC <- err
+		return
+	}
+	if !*isSynced {
+		errC <- fmt.Errorf("clique not synced")
+		return
+	}
+
+	w.logger.Info("alephium watcher started", zap.String("version", nodeInfo.BuildInfo.ReleaseVersion), zap.Uint32("fromEventIndex", w.fromEventIndex))
+
+	eventsC := make(chan []*EventsPerIndex)
+	go w.fetchEvents(ctx, w.fromEventIndex, errC, eventsC)
```

Comment:
To avoid missing events, the watcher reads the latest event index from the db each time it starts running, so we don't need to update the `w.fromEventIndex`. The function `fetchEvents` will fetch events from `w.fromEventIndex` and the next event index will be updated after each request: https://github.com/alephium/wormhole-fork/pull/63/files#diff-690c49dc60e230bb5ded0e3173e669e1ae433cb23c60ce84e16cdac5570fa65fR121

Reply:
Got it, thanks!